# anya's keywords

In [1]:
from tqdm import tqdm
import os
import json

In [2]:
import pandas as pd
import numpy as np

from google.colab import drive
drive.mount('/gdrive')


Mounted at /gdrive


In [3]:
with open("/gdrive/My Drive/nnproj/all_keywords.json") as f:
    kws = json.loads(f.read())

In [4]:
kws.keys() # ids of the hotels

dict_keys(['AVwc252WIN2L1WUfpqLP', 'AVwdOclqIN2L1WUfti38', 'AVwePiAX_7pvs4fzBSAl', 'AVwdUKE3_7pvs4fz4lE-', 'AVweQyWNkufWRAb5BZeg', 'AVweQDZpIN2L1WUf3TA9', 'AVwdQ0dy_7pvs4fz4BqX', 'AWE7dKHFIxWefVJwyLgA', 'AVwcfFGw_7pvs4fzvk4S', 'AVwdbM6R_7pvs4fz5wdc', 'AVwemBCCkufWRAb5EZCs', 'AWE2CYSsIxWefVJwyDrn', 'AV-TGsFqRxPSIh2RmVF-', 'AVwct7BzIN2L1WUfoHaP', 'AWKpI38Q3-Khe5l_hxUR', 'AWKlQWNQa4HuVbedPUSi', 'AVzX1GbfFcQ3k02bB73-', 'AWEhoizBa4HuVbedNJo_', 'AV-TGwqGIxWefVJwr5Ji', 'AV3O9oECIxWefVJwjraT', 'AWNSHai33-Khe5l_ifzh', 'AWE7h8u53-Khe5l_f4mi', 'AV1eTXQF3-Khe5l_NF6Q', 'AWE7h7Cta4HuVbedNeQt', 'AWE2E2OG3-Khe5l_fwmT', 'AWKpYhaja4HuVbedPVes', 'AWKkrZYpa4HuVbedPT1I', 'AVzOLyE1FcQ3k02bBr9S', 'AWE2BgHcIxWefVJwyDYU', 'AVzX37zs-gnIPe8DTHi5', 'AVzYkxFXFcQ3k02bCKjD', 'AV-TGuWBa4HuVbedGcJK', 'AWKr8UBHa4HuVbedPWl2', 'AWKuEeWka4HuVbedPXff', 'AWE7he303-Khe5l_f4kC', 'AWEw3H-UIxWefVJwyAZ0', 'AV-TGwK3IxWefVJwr5IJ', 'AWUPNmq6IxWefVJw3c71', 'AWE1_Z8-RxPSIh2Rsaks', 'AWKoW_tYa4HuVbedPVUg', 'AVwdoQs4IN2L1WUfxtHH', 'AVwd

In [5]:
len(kws)

2197

In [6]:
kws['AVwc252WIN2L1WUfpqLP'] 

{'neg': [],
 'pos': ['Amazing place',
  'some top notch places',
  'the highest rated tennis resorts',
  'an amazing experience',
  'my new favorite places',
  'perfect experience',
  'perfect villas',
  'impressive program',
  'amazing experience']}

In [9]:
kws["AVwdOclqIN2L1WUfti38"]

{'neg': ['nothing'], 'pos': ['a great location', 'super nice staff']}

In [ ]:
# list(kws.keys())[2]

In [11]:
kws["AVwePiAX_7pvs4fzBSAl"]

{'neg': [],
 'pos': ['Very friendly staff',
  'great free breakfast items',
  'a nice hotel',
  'great staff',
  'young children',
  'Beautiful property',
  'excellent customer service']}

# кластеризация не нужна

clustering is not an optimal choice, as we need to cluster inside positive and negative sets inside 1 hotel, and this is not enough data for the clustering algorithms.

## tfidf sorting for keywords ONLY WHERE THERE ARE A LOT OF THEM. selecting top 20.

много это больше 20 (цифра из головы)



----------

In [ ]:
cpos = set()
cneg = set()

for key in kws.keys():
  if len(kws[key]['neg']) > 20:
      cneg.add(key)

  if len(kws[key]['pos']) > 20:
      cpos.add(key)


## те отели у которых количество негативных отзывов > 20, это те же отели у которых количество позитивных отзывов > 20


In [21]:
len(cneg), len(cpos)

(74, 245)

In [30]:
len(cneg & cpos), len(cneg | cpos)

(74, 245)

# for those 245 hotels with a lot of keywords

example

In [40]:
len(kws["AV48ic48a4HuVbedAUNM"]['neg']), len(kws["AV48ic48a4HuVbedAUNM"]['pos'])


(47, 119)

In [44]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
stop = stopwords.words('english')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [42]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [116]:
biglist = []

чтобы не испортить tfidf'ом имеющиеся N-grammы склеиваю их нижним подчеркиванием

In [117]:
for key in kws:
  ptemp = []
  for el in kws[key]["pos"]:
    ptemp.append(el.replace(" ", "_"))

  ntemp = []
  for el in kws[key]["neg"]:
    ntemp.append(el.replace(" ", "_"))

  temp = {
      "hotel_id": key,
      "pos": " ".join(ptemp),
      "neg": " ".join(ntemp),
  }
  biglist.append(temp)

In [118]:
dfall = pd.DataFrame(biglist)

In [120]:
mask = dfall['hotel_id'].isin(cpos)

датафрейм отелей у которых позитивных ключевых слов больше 20

In [121]:
df_popular = dfall.loc[mask]

In [122]:
df_popular.head()

hotel_id                                                pos  \
7   AWE7dKHFIxWefVJwyLgA  the_perfect_price Well_sir Gaylord_Hotel Good_...   
11  AWE2CYSsIxWefVJwyDrn  Awesome_hotel the_really_cool_room the_other_a...   
12  AV-TGsFqRxPSIh2RmVF-  great_location Good_location Very_pleasant_exp...   
14  AWKpI38Q3-Khe5l_hxUR  a_very_good_price Best_gambling best_comps exc...   
15  AWKlQWNQa4HuVbedPUSi  good_food its_fun What_an_awesome_experience 1...   

                                                  neg  
7                                          small_room  
11  The_room_rate a_Booking.com_rate no_explanatio...  
12  a_disappointment disgrace air_con nice_too_sma...  
14           resort_fees nothing nothing good_buffett  
15  the_counter sleep old_room no_roulette the_bat...

### tfidf on all pos keywords of popular hotels

In [130]:
vectorizer=TfidfVectorizer(max_features=1000,stop_words=stop)
X = vectorizer.fit_transform(df_popular["pos"])

In [131]:
X.shape

(245, 1000)

In [132]:
X = X.toarray()

feats = vectorizer.get_feature_names_out()

In [133]:
n = 20
kws_ind = (-X).argsort()[:,:n] # top n key word for every hotel

In [134]:
kws_ind, kws_ind.shape

(array([[612, 132, 951, ..., 679, 678, 660],
        [587,  68, 307, ..., 682, 666, 667],
        [949, 528, 370, ..., 662, 663, 664],
        ...,
        [124, 858, 634, ..., 776, 304, 735],
        [661, 823, 219, ..., 608,  82,  83],
        [502, 514, 150, ..., 172, 417,  80]]), (245, 20))

In [ ]:
forcolumn = []
for el in feats[kws_ind]:
  forcolumn.append(" ".join(el))

df_popular["pos_top20"] = forcolumn

In [187]:
df_popular.head()

hotel_id                                                pos  \
7   AWE7dKHFIxWefVJwyLgA  the_perfect_price Well_sir Gaylord_Hotel Good_...   
11  AWE2CYSsIxWefVJwyDrn  Awesome_hotel the_really_cool_room the_other_a...   
12  AV-TGsFqRxPSIh2RmVF-  great_location Good_location Very_pleasant_exp...   
14  AWKpI38Q3-Khe5l_hxUR  a_very_good_price Best_gambling best_comps exc...   
15  AWKlQWNQa4HuVbedPUSi  good_food its_fun What_an_awesome_experience 1...   

                                                  neg  \
7                                          small_room   
11  The_room_rate a_Booking.com_rate no_explanatio...   
12  a_disappointment disgrace air_con nice_too_sma...   
14           resort_fees nothing nothing good_buffett   
15  the_counter sleep old_room no_roulette the_bat...   

                                            pos_top20  
7   great_staff a_nice_area very_good_experience e...  
11  great_happy_hour a_great_bonus beautiful_bar v...  
12  very_friendly_staff good_location clean_linens...  
14  its_price a_very_good_price a_good_shower grea...  
15  good_buffet a_great_view good_gym good_locatio...

### tfidf on all NEGATIVE keywords of popular hotels

taking from the previous df only those hotels that have a lot of negative kws

In [188]:
mask = dfall['hotel_id'].isin(cneg)
neg_popular_df = df_popular.loc[mask] # so that pos_top20 was there

In [189]:
vectorizer=TfidfVectorizer(max_features=1000,stop_words=stop)
X = vectorizer.fit_transform(neg_popular_df["neg"])

In [190]:
X.shape

(74, 1000)

In [191]:
X = X.toarray()

feats = vectorizer.get_feature_names_out()

In [192]:
n = 20
kws_ind = (-X).argsort()[:,:n] # top n key word for every hotel

In [193]:
kws_ind, kws_ind.shape

(array([[695, 506, 897, ..., 156, 901, 253],
        [142, 938, 916, ...,  22, 104, 497],
        [279, 689,  20, ..., 882, 649, 389],
        ...,
        [839,  48, 332, ..., 338,  36, 571],
        [581, 670, 390, ..., 980,  93, 279],
        [279,   6, 136, ..., 402, 797, 186]]), (74, 20))

In [194]:
forcolumn = []
for el in feats[kws_ind]:
  forcolumn.append(" ".join(el))

neg_popular_df["neg_top20"] = forcolumn

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


In [195]:
neg_popular_df.head()

hotel_id                                                pos  \
15  AWKlQWNQa4HuVbedPUSi  good_food its_fun What_an_awesome_experience 1...   
26  AWKkrZYpa4HuVbedPT1I  the_best_spot All_amenities car_park Great_loc...   
29  AVzX37zs-gnIPe8DTHi5  Wonderful_hotel the_higher_rate higher_rates w...   
36  AV-TGwK3IxWefVJwr5IJ  Fantastic_location Fun_location very_nice_hote...   
46  AVzOLf4aFcQ3k02bBrRN  the_Best_Western_Brand Clean_and_comfortable_r...   

                                                  neg  \
15  the_counter sleep old_room no_roulette the_bat...   
26  no_extra_cost my_booking_ticket joke a_bad_exp...   
29  an_anniversary_stay no_free_lunch too_much_tro...   
36  no_non-dairy_options no_windows no_restaurant ...   
46  nothing The_only_trace hash an_overpowering_st...   

                                            pos_top20  \
15  good_buffet a_great_view good_gym good_locatio...   
26  great_location no_hot_water friendly_staff com...   
29  important_comments great_views south_beach bea...   
36  great_location happy_hour outstanding_staff ex...   
46  the_best_western best_western friendly_staff b...   

                                            neg_top20  
15  the_drain satisfactory_buffet ugly_people toug...  
26  no_casino vet_bad_service unhappy_guest a_bad_...  
29  nothing the_disappointing_experience a_report ...  
36  poor_breakfast no_windows no_non slow_valet no...  
46  this_minor_problem very_poor_insulation total_...

# final json for anya

In [222]:
final = {}

In [223]:
neghotels = list(neg_popular_df["hotel_id"])


delete _ from n gramms

In [224]:
for hotel_id in neghotels:
    neg_top = list(neg_popular_df.loc[neg_popular_df['hotel_id'] == hotel_id].neg_top20)[0]
    pos_top = list(neg_popular_df.loc[neg_popular_df['hotel_id'] == hotel_id].pos_top20)[0]

    neg_to_normal = []
    for el in neg_top.split():
        neg_to_normal.append(el.replace("_", " "))
    
    pos_to_normal = []
    for el in pos_top.split():
        pos_to_normal.append(el.replace("_", " "))
    
    if hotel_id in final:
      raise Exception("shouldnt be")

    final[hotel_id] = {"neg": neg_to_normal,
                       "pos": pos_to_normal
                       }


now add popular hotels with a lot of positive kws and small num of negative

In [225]:
mask = dfall['hotel_id'].isin(cneg)
without_neg_popular = df_popular.loc[~mask]

In [226]:
without_neg_popular.shape

(171, 4)

In [229]:
hotels = list(without_neg_popular["hotel_id"])
len(hotels)

171

In [230]:
without_neg_popular.head(3)

hotel_id                                                pos  \
7   AWE7dKHFIxWefVJwyLgA  the_perfect_price Well_sir Gaylord_Hotel Good_...   
11  AWE2CYSsIxWefVJwyDrn  Awesome_hotel the_really_cool_room the_other_a...   
12  AV-TGsFqRxPSIh2RmVF-  great_location Good_location Very_pleasant_exp...   

                                                  neg  \
7                                          small_room   
11  The_room_rate a_Booking.com_rate no_explanatio...   
12  a_disappointment disgrace air_con nice_too_sma...   

                                            pos_top20  
7   great_staff a_nice_area very_good_experience e...  
11  great_happy_hour a_great_bonus beautiful_bar v...  
12  very_friendly_staff good_location clean_linens...

In [231]:
for hotel_id in hotels:
    neg_top = list(without_neg_popular.loc[without_neg_popular['hotel_id'] == hotel_id].neg)[0]
    pos_top = list(without_neg_popular.loc[without_neg_popular['hotel_id'] == hotel_id].pos_top20)[0]

    neg_to_normal = []
    for el in neg_top.split():
        neg_to_normal.append(el.replace("_", " "))
    
    pos_to_normal = []
    for el in pos_top.split():
        pos_to_normal.append(el.replace("_", " "))
    
    if hotel_id in final:
      raise Exception("shouldnt be")

    final[hotel_id] = {"neg": neg_to_normal,
                       "pos": pos_to_normal
                       }

    

In [232]:
len(final)

245

In [234]:
len(kws)

2197

заменяю на новые данные при совпадении hotel id

In [235]:
kws.update(final)

In [237]:
kws["AWE7dKHFIxWefVJwyLgA"]

{'neg': ['small room'],
 'pos': ['great staff',
  'a nice area',
  'very good experience',
  'excellent experience',
  'nice views',
  'fantastic location',
  'nice breakfast',
  'comfortable rooms',
  'great value',
  'a great stay',
  'comfortable bed',
  'clean room',
  'friendly staff',
  'lovely hotel',
  'internet access',
  'magazine st',
  'its price',
  'lovely staff',
  'lovely rooms',
  'key service']}

In [238]:
with open("/gdrive/My Drive/nnproj/all_keywords_with_top20.json", "w") as f:
    json.dump(kws, f)
    

In [239]:
# with open("/gdrive/My Drive/nnproj/all_keywords_with_top20.json") as f:
#     kws = json.loads(f.read())